### PR TITLE
Tweak std::marker docs

### LINKED ISSUE #29361: API Docs: marker
part of https://github.com/rust-lang/rust/issues/29329

http://doc.rust-lang.org/std/marker/

### PR DESCRIPTION
**Pull request by kmcallister**
Fixes #29361.

r? @GuillaumeGomez

### CODE REVIEW

File: src/libcore/marker.rs [hunk 1]
```diff
@@ -60,13 +76,16 @@ pub trait Sized {
 }
 
 /// Types that can be "unsized" to a dynamically sized type.
+///
+/// For example, the sized array type `[i8; 2]` implements `Unsize<[u8]>` and
```
Comment (sfackler) [inline src/libcore/marker.rs]: It might be worth briefly talking about how `Unsize` is paired with `CoerceUnsized` and what it's used for here.

Comment (kmcallister) [inline src/libcore/marker.rs]: Yeah, that's something I don't understand particularly well. Do you have a reference?

Comment (sfackler) [inline src/libcore/marker.rs]: I think the RFC should still mostly reflect the current reality but I'm not super familiar with DST coercions either: https://github.com/nrc/rfcs/blob/4b99006cac67f7ad89c670bcb351d32bc0d5d33f/text/0000-dst-coercion.md

File: src/libcore/marker.rs [hunk 2]
```diff
@@ -38,10 +53,11 @@ impl<T: ?Sized> !Send for *const T { }
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T: ?Sized> !Send for *mut T { }
 
-/// Types with a constant size known at compile-time.
+/// Types with a constant size known at compile time.
 ///
-/// All type parameters which can be bounded have an implicit bound of `Sized`. The special syntax
-/// `?Sized` can be used to remove this bound if it is not appropriate.
+/// All type parameters which can be bounded have an implicit bound of `Sized`.
```
Comment (durka) [inline src/libcore/marker.rs]: Are there any type parameters which cannot be bounded?

Comment (kmcallister) [inline src/libcore/marker.rs]: I'm not sure. The only case I know of where the `Sized` bound isn't automatic is the implicit `Self` parameter of a trait definition.

**Comment by frewsxcv** [comment]
```
Linkcheck stage2 (x86_64-unknown-linux-gnu)
core/marker/trait.Sync.html:98: broken link - core/mem/fn.transmute.html
thread 'main' panicked at 'found some broken links', src/tools/linkchecker/main.rs:54
```

File: src/libcore/marker.rs [hunk 3]
```diff
@@ -126,107 +193,111 @@ pub trait Unsize<T: ?Sized> {
...
+/// [cell]: ../cell/struct.Cell.html
+/// [refcell]: ../cell/struct.RefCell.html
+/// [rc]: ../../std/rc/struct.Rc.html
+/// [arc]: ../../std/sync/struct.Arc.html
+/// [atomic data types]: ../sync/atomic/index.html
+/// [mutex]: ../../std/sync/struct.Mutex.html
+/// [rwlock]: ../../std/sync/struct.RwLock.html
+/// [unsafecell]: ../cell/struct.UnsafeCell.html
+/// [ub]: ../../reference.html#behavior-considered-undefined
+/// [transmute]: ../mem/fn.transmute.html
```
Comment (frewsxcv) [inline src/libcore/marker.rs]: `core::mem::transmute` doesn't exist, which might make this problematic?

Comment (GuillaumeGomez) [inline src/libcore/marker.rs]: Add "../std".

**Comment by kmcallister** [comment]
Addressed review comments. Thanks everyone.

**Comment by GuillaumeGomez** [comment]
Please squash your commits! Once done, I'd prefer to wait that #36363 has been merged before r+ yours to avoid conflicts.

**Comment by kmcallister** [comment]
Yeah, that's going to conflict. I'll fix my PR once it lands.

**Comment by kmcallister** [comment]
I resolved conflicts and rebased.

File: src/libcore/marker.rs [hunk 4]
```diff
@@ -455,21 +570,25 @@ mod impls {
...
+/// is often called the "[parametricity property][param]". Despite the
+/// use of `Reflect`, Rust lacks true parametricity because a generic
+/// function can, at the very least, call [`mem::size_of`][size_of]
+/// without employing any trait bounds whatsoever.
+///
+/// [any]: ../any/trait.Any.html
+/// [typeid]: ../any/struct.TypeId.html
+/// [param]: http://en.wikipedia.org/wiki/Parametricity
+/// [size_of]: ../mem/fn.size_of.html
+///
```
Comment (GuillaumeGomez) [inline src/libcore/marker.rs]: This line is useless.

File: src/libcore/marker.rs [hunk 5]
```diff
@@ -408,21 +520,21 @@ macro_rules! impls{
...
+/// `T`. This has bearing on the Rust compiler's [drop check]
+/// analysis.
 ///
 /// If your struct does not in fact *own* the data of type `T`, it is
 /// better to use a reference type, like `PhantomData<&'a T>`
 /// (ideally) or `PhantomData<*const T>` (if no lifetime applies), so
 /// as not to indicate ownership.
+///
+/// [drop check]: ../../nomicon/dropck.html
+///
```
Comment (GuillaumeGomez) [inline src/libcore/marker.rs]: This line is useless.

Comment (kmcallister) [inline src/libcore/marker.rs]: I saw that some of the existing docs have it. I think it makes things more readable in the source code. If you feel very strongly I'll remove it, though.

Comment (kmcallister) [inline src/libcore/marker.rs]: Fixed.

File: src/libcore/marker.rs (hunk 4 above)
Comment (kmcallister) [inline src/libcore/marker.rs]: Fixed.

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r+ rollup

**Pull request merged**